### PR TITLE
Add broadcast

### PR DESCRIPTION
**Pull request by itabrl**
# PR Details
## Description

Broadcast Vehicle Status And Intent (Class C and Class D)

## Motivation and Context

As CARMA Platform on a vehicle entering an unsignalized intersection, vehicle broadcast status as an Entering Vehicle so that other vehicles can estimate and plan their stopping trajectories.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [x] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: sci_strategic_plugin/include/sci_strategic_plugin.h [hunk 1]
```diff
@@ -247,6 +247,11 @@ class SCIStrategicPlugin
    *
    */
   void caseThreeSpeedProfile(double stop_dist, double current_speed, double stop_time, std::vector<double>* float_metadata_list) const;
+
+
+  cav_msgs::MobilityOperation generateMobilityOperation();
```
Comment (saina-ramyar) [inline sci_strategic_plugin/include/sci_strategic_plugin.h:252]: add function description

File: sci_strategic_plugin/include/sci_strategic_plugin.h [hunk 2]
```diff
@@ -247,6 +247,11 @@ class SCIStrategicPlugin
...
+
+  std::vector<uint8_t> getMsgId(const ros::Time now);
```
Comment (saina-ramyar) [inline sci_strategic_plugin/include/sci_strategic_plugin.h:254]: add function description

File: sci_strategic_plugin/include/sci_strategic_plugin.h [hunk 3]
```diff
@@ -269,10 +274,15 @@ class SCIStrategicPlugin
 
   bool approaching_stop_controlled_interction_ = false;
 
+  double vehicle_status_generation_frequency_;
```
Comment (saina-ramyar) [inline sci_strategic_plugin/include/sci_strategic_plugin.h:277]: needs to be initilized

File: sci_strategic_plugin/include/sci_strategic_plugin_config.h [hunk 4]
```diff
@@ -57,5 +57,7 @@ struct SCIStrategicPluginConfig
 
   //! License plate of the vehicle.
   std::string vehicle_id = "default_id";
+
+  double vehicle_status_generation_frequency_;
```
Comment (saina-ramyar) [inline sci_strategic_plugin/include/sci_strategic_plugin_config.h:61]: parameter should have an initial value

File: sci_strategic_plugin/src/main.cpp [hunk 5]
```diff
@@ -50,6 +50,7 @@ int main(int argc, char** argv)
   pnh.param<std::string>("lane_following_plugin_name",       config.lane_following_plugin_name, config.lane_following_plugin_name);
   pnh.param<std::string>("intersection_transit_plugin_name", config.intersection_transit_plugin_name, config.intersection_transit_plugin_name);
   pnh.getParam("/vehicle_id", config.vehicle_id);
+  pnh.param<double>("vehicle_status_generation_frequency_", config.vehicle_status_generation_frequency_, 10.0);
```
Comment (saina-ramyar) [inline sci_strategic_plugin/src/main.cpp:53]: param value should be set in config/parameters.taml, so they can be changed during tests

File: sci_strategic_plugin/src/sci_strategic_plugin.cpp [hunk 6]
```diff
@@ -394,4 +397,43 @@ double SCIStrategicPlugin::findSpeedLimit(const lanelet::ConstLanelet& llt) cons
   }
 }
 
+cav_msgs::MobilityOperation SCIStrategicPlugin::generateMobilityOperation()
+{
+    cav_msgs::MobilityOperation mo_;
+    mo_.header.timestamp = ros::Time::now().toNSec();
```
Comment (saina-ramyar) [inline sci_strategic_plugin/src/sci_strategic_plugin.cpp:403]: timestamp should be in milisecond

File: sci_strategic_plugin/src/sci_strategic_plugin.cpp [hunk 7]
```diff
@@ -394,4 +397,43 @@ double SCIStrategicPlugin::findSpeedLimit(const lanelet::ConstLanelet& llt) cons
...
+    mo_.header.sender_bsm_id = std::string(reinterpret_cast<const char*>(getMsgId(ros::Time::now())[0]), getMsgId(ros::Time::now()).size());
```
Comment (saina-ramyar) [inline sci_strategic_plugin/src/sci_strategic_plugin.cpp:404]: are you creating bsm id here? The bsm id should be extracted from bsm msgs

File: sci_strategic_plugin/src/sci_strategic_plugin.cpp [hunk 8]
```diff
@@ -394,4 +397,43 @@ double SCIStrategicPlugin::findSpeedLimit(const lanelet::ConstLanelet& llt) cons
...
+
+    double vehicle_acceleration_limit_ = config_.vehicle_accel_limit * config_.vehicle_accel_limit_multiplier;
+    double vehicle_deceleration_limit_ = -1 * config_.vehicle_decel_limit * config_.vehicle_decel_limit_multiplier;
+
+    mo_.strategy_params = "intersection_box_flag, acceleration_limit, deceleration_limit," + std::to_string(flag) + "," + std::to_string(vehicle_acceleration_limit_) + "," + std::to_string(vehicle_deceleration_limit_);
+
+    mobility_operation_pub.publish(mo_);
+}
+
+std::vector<uint8_t> SCIStrategicPlugin::getMsgId(const ros::Time now)
```
Comment (saina-ramyar) [inline sci_strategic_plugin/src/sci_strategic_plugin.cpp:416]: As mentioned above, I dont think we should generate bsm id ourself. We need to subscribe to the bsm msg and extract the id from that. The bsm id is how the status and bsm messages can be matched,

**Review by saina-ramyar — COMMENTED** [review]
Also please add unit tests and ros test.

File: sci_strategic_plugin/src/sci_strategic_plugin.cpp (hunk 6 above)
Comment (itabrl) [reply sci_strategic_plugin/src/sci_strategic_plugin.cpp:408]: This matches the code from the port_drayage_plugin. should we discuss this?

File: sci_strategic_plugin/src/sci_strategic_plugin.cpp [hunk 9]
```diff
@@ -394,4 +402,23 @@ double SCIStrategicPlugin::findSpeedLimit(const lanelet::ConstLanelet& llt) cons
   }
 }
 
-}  // namespace SCI_strategic_plugin
\ No newline at end of file
+void SCIStrategicPlugin::generateMobilityOperation()
+{
+    cav_msgs::MobilityOperation mo_;
+    mo_.header.timestamp = ros::Time::now().toNSec()/1000;
```
Comment (saina-ramyar) [inline sci_strategic_plugin/src/sci_strategic_plugin.cpp:408]: ms timestamp is: ros::Time::now().toNSec() * 1000000;

**Pull request merged**